money-aspectj/src/test/scala/com/comcast/money/aspectj/TraceAspectSpec.scala

```diff
@@ -76,6 +79,12 @@ class TraceAspectSpec extends WordSpec
     throw new RuntimeException("not-ignored")
   }
 
+  @Traced(
```

Comment:
As opposed to adding an async flag, with aspectj anyway we can add a different joinpoint / pointcut on functions with annotations `@Trace` that return a promise (I think).  Have you tried that?  Unsure off the top of my head if that would work

Reply:
That would probably work although might require some effort to avoid ambiguity between the existing pointcut.  The problem with that is that it locks it to supporting only specific types and there is a fairly wide ecosystem of these types out there since Java didn't add one offering a callback until Java 8.  `Promise[T]` is quite specific to Scala and its uncommon to see in other Java libraries.

Reply:
Right, I am unclear if it would work.  Can we have a test where, if someone sets async to be true, on something that is not async, that everything still works?

Imagine

```
@Traced(value="sync", async=true)
def plusOne(x: Int): Int = x + 1
```

Reply:
Agreed, the tests are a little anemic right now.  This PR is more a spike for the purpose of discussion.

I might need to approach the unit tests a little differently as they should validate that log records aren't written until the promise/future is complete and the current helper methods don't really lend themselves to that, short of waiting for a timeout to expire.  I'm thinking of mocking the Tracer instead.

Reply:
The main test that I would like to try out is nested futures (futures with flatmaps).

When I worked with scala Futures in the past, the onComplete of a future was executed in odd ways.

Imagine you have 3 levels of functions / nested futures via flatmaps and maps.  You create child spans at each level, so 3 levels deep in a span tree.  The onComplete and context that was passed was "off".

Sorry I don't have more information, I have to check out your PR and try and re-create the issue I saw

Reply:
You mean something like this?

```scala
@Traced(value = "asyncGetHello", async = true)
def asyncGetHello(): Future[String] = Future { "hello" }
@Traced(value = "asyncGetWorld", async = true)
def asyncGetWorld(): Future[String] = Future { "world" }

@Traced(value = "someAsyncMethod", async = true)
def someAsyncMethod(): Future[String] = {
    for {
        hello <- asyncGetHello()
        world <- asyncGetWorld()
    } yield s"$hello $world!"
}
```

Reply:
I think I see what you mean.  While the above does produce three spans as expected, the parent of span "asyncGetWorld" is not "someAsyncMethod".  Is that the behavior that you experienced?

Reply:
So, assuming that this is the problem to which you were referring, I've thought of a way in which it can be resolved, although it's a little heavy-handed.  The issue is that we do not control the thread on which `asyncGetWorld` is invoked.  At best we can hand a custom `ExecutionContext`, such as the `TraceFriendlyExecutionContextExecutor` already in `money-core`.  The problem is also that the thread on which `ExecutionContext.execute` is invoked is the thread on which the `Future` happened to complete, which is not the thread that we control and it's too late to capture a span.  The best we can do here is to have `TraceFriendlyExecutionContextExecutor` capture a span as soon as it is created and to apply that span on every `Runnable` submitted to `execute`.  But this requires that a new `TraceFriendlyExecutionContextExecutor` be created at the beginning of every single async method and prevents importing a common executor.

```scala
@Traced(value = "asyncGetHello", async = true)
def asyncGetHello(): Future[String] = Future { "hello" }
@Traced(value = "asyncGetWorld", async = true)
def asyncGetWorld(): Future[String] = Future { "world" }

@Traced(value = "someAsyncMethod", async = true)
def someAsyncMethod(): Future[String] = {
    implicit val ec: ExecutionContext = TraceFriendlyExecutionContext(ExecutionContext.global, SpanLocal.current)
    for {
        hello <- asyncGetHello()
        world <- asyncGetWorld()
    } yield s"$hello $world!"
}
``` 

We could probably make that less verbose, but it still puts a lot of onus on the developer to get right.

Alternately, I have tested an aspect which intercepts all calls to `Future` methods which accept an `ExecutionContext` within a method annotated with `@Traced` which wraps the `ExecutionContext` with one that immediately captures the span.  This seems to work pretty well, but as I mentioned, is quite heavy handed.

A third option might be to have an aspect that intercepts all calls to any method that can return a `Future` within a method annotated with `@Traced` and wrap the `Future` itself.  I don't think that's much different and may prove more problematic for code that might extend from `Future` directly.

Reply:
In the version 0.8.x codebase, I had written a `TracedFuture` that handled this well.  I abandoned it, maybe I shouldn't have.  I will dig that up.  There is also a unit test that tests out the proper behavior

Reply:
Here is the traced future stuff - https://github.com/Comcast/money/blob/0.8.x/money-core/src/main/scala/com/comcast/money/concurrent/Futures.scala

Reply:
Here are the unit tests that illustrate the problem ....

https://github.com/Comcast/money/blob/0.8.x/money-core/src/test/scala/com/comcast/money/concurrent/FuturesSpec.scala

Reply:
I would be interested to test out the second approach you mentioned.  The code I had extended Futures, but iirc, not all use cases were cleanly supported.  Plus, we have this TracedFuture thing.

Reply:
I've pushed a second branch to my fork of the repo that contains this aspect and the simple test above.

https://github.com/HaloFour/money/blob/async-tracing-future-aspect/money-aspectj/src/main/scala/com/comcast/money/aspectj/FutureAspect.scala

Reply:
I've been playing around a little with the `TracedFuture[T]` code and unit tests that you linked, and I think it runs into a fundamental problem when it comes to the lifespan of the span.  In your example you are stopping the span as soon as the `TracedFuture[T]` completes, but that happens before any composed operations are executed, so you'd lose recording any data to the span from those callbacks.  But if you leave the span open there is no way to know when the span should be stopped.  You'd need a kind of terminal operator which the developer would be required to call at the end of their composition.

Rather than going to that trouble it might be easier to expose a helper method in which you'd wrap all of the future logic:

```scala
object Futures {
  private val tracer = Money.Environment.tracer

  def traceFuture[T](key: String)(f: (ExecutionContext) => Future[T])(implicit executionContext: ExecutionContext): Future[T] = {
    val span = SpanLocal.current match {
      case Some(existingSpan) => tracer.spanFactory.childSpan(key, existingSpan)
      case _ => tracer.spanFactory.newSpan(key)
    }

    val future = f(wrap(executionContext, span, captureMdc()))

    future onComplete {
      case Success(_) => span.stop(true)
      case _ => span.stop(false)
    }

    future
  }
}
```

then:

```scala
traceFuture("root") { implicit ec =>
  tracer.record("begin", "root")
  println(s"\r\n; begin root ${SpanLocal.current}")

  traceFuture("nested") { _ =>
    tracer.record("begin", "nested")
    println(s"\r\n; begin nested ${SpanLocal.current}")
    Future { Some(456) }
  }.flatMap {
    case Some(v) =>
      tracer.record("flatMap", "nested")
      println(s"\r\n; flatMap nested ${SpanLocal.current}")
      Future {
        v
      }
  }.map {
    _ =>
      tracer.record("map", "nested")
      println(s"\r\n; map nested ${SpanLocal.current}")
      SpanLocal.current.get
  }
}.map { child =>
  tracer.record("flatMap", "root")
  println(s"\r\n; flatMap root ${SpanLocal.current}")
  child
}.map {
  s =>
    tracer.record("map", "root")
    println(s"\r\n; map root ${SpanLocal.current}")
    (SpanLocal.current.get, s)
}
```

Reply:
Right, I recall now that we did not bring the TracedFuture forward due to stopping spans.

The old 0.8.x tracing was based on actors, and spans would "hang out" for a bit before being released.  This allowed the span to be stopped multiple times, so the TracedFuture stuff worked ok.

We do not do that in the current 0.9.x branch.

I was thinking of using a monad transformer based around StateT.  The cedi guys did something similar in cedi-dtrace; however there design was based on ReaderT, which did not allow flexible modification of the Span in flight.

https://github.com/ccadllc/cedi-dtrace

Would love to build on the cedi stuff, use StateT instead.

I wonder if we had that, would the trace aspect stuff become easier?  Like, we could wrap a Future in a StateT at certain points, and then pull it out of StateT?

Reply:
I'll have to spend some time grokking over cedi and cats to get a better understanding of it.  I'm a bit of a Scala neophyte at the moment.  My concern is what this would look like over in Java-land where the Scala sugar would be absent.  How much onus does this put on the developer to explicitly spell out their tracing?  How would it work over other types?

Most of the complexity of the additions to `TraceAspect` here are in my attempt to be future-agnostic, hoping that it would eventually support some of the several abstractions ecosystem, particularly Java 8 `CompletionStage`, Java 9 `Flow`, RxJava and Spring Reactor, all of which work a little differently.  I reached for Scala `Future` here because it was already there.  It works well enough for that purpose, as long as you don't compose of the future within the annotated method.  The implementation details would be different for each abstraction, though.

I was able to greatly simplify the [`FutureAspect`](https://github.com/HaloFour/money/blob/async-tracing-future-aspect/money-aspectj/src/main/scala/com/comcast/money/aspectj/FutureAspect.scala) I was playing with in the other branch so that it only required two pointcuts and a single advice.  It works well with or without the `@Trace` aspect.  I need to write up some tests to prove it out.

Reply:
I think if you can manage to get those old trace unit tests working properly, we should be good.

I really like the idea of extending this to other async metaphors too, that is great.

Sorry about the slow turnaround on this stuff.  I have been swamped lately.  Busy clearing things off my plate so I can dive into this deeper.

Reply:
No worries.  I've been pretty swamped too.

I've taken your tests and made the majority of them work using TraceAspect and FutureAspect.  The major ones that are still failing are those that use the `onSuccess` and `onFailure` methods as they don't return new `Future` instances which creates a race condition between the callback logging to the span and the span being closed.  I don't have a good answer as to how to solve that yet.

Reply:
I wonder if we can just have a different type of `Span` created different than `CoreSpan`...one that can accept multiple stops.

In a really high throughput system if you sample EVERY request, that can be tough.  One of the biggest misses so far with money is the lack of sampling.  Basic sampling strategies are so simple to implement I kick myself for not doing it years ago.  If we had some level of sane sampling, we wouldn't have to worry too much about memory pressure.

Imo, we can partially close a span on the first stop, and then kinda wait a little before releasing the span info for handling.  With a sane sampling rate that is high (like 100 samples per second), those spans could easily hang out for a minute or so.

Reply:
Also, I think your example is odd.  Seems like you have a future that yields a future as a value.  That is dumb (not calling you dumb, but that doesn't seem like a good use case).  I am curious if some of the contrived tests I wrote are things that people would never write?

Certainly `.flatMap { x => Future { x + 1 } }` is valid

But `Future { Future { x }  }` is not

Reply:
Which example is that?  I'm trying to duplicate (or at least reinterpret) the unit tests you had around traced futures.  Some of the examples are contrived intentionally but I also wouldn't discount my flubbing things up.  I've not had the luxury of dedicating my attention to this recently.

Reply:
I was playing with going down a path of having `TraceAspect` replace the existing `Span` with a wrapper `AsyncSpan` which internally would use an `AtomicInteger` as a reference counter.  Invoking a terminal method on `Future` like `onComplete` would bump that counter up.  When that callback would complete it would call `stop`, which would decrease the counter, as would when the instrumented future completed.  Only when the counter reached zero would the span actually be stopped.  I don't think that's quite what you were talking about, though.

Reply:
That’s a great idea!

Reply:
I was thinking that perhaps my old TraceAspect unit tests had test cases that were invalid.